misc/bazel/internal/git_lfs_probe.py

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python3
+
+"""
+Probe lfs files.
+For each source file provided as output, this will print:
+* "local", if the source file is not an LFS pointer
+* the sha256 hash, a space character and a transient download link obtained via the LFS protocol otherwise
+"""
+
+import sys
+import pathlib
+import subprocess
+import os
+import shutil
+import json
+import urllib.request
+from urllib.parse import urlparse
+import re
+
+sources = [pathlib.Path(arg).resolve() for arg in sys.argv[1:]]
+source_dir = pathlib.Path(os.path.commonpath(src.parent for src in sources))
+source_dir = subprocess.check_output(["git", "rev-parse", "--show-toplevel"], cwd=source_dir, text=True).strip()
+
+
+def get_endpoint():
+    lfs_env = subprocess.check_output(["git", "lfs", "env"], text=True, cwd=source_dir)
+    endpoint = ssh_server = ssh_path = None
+    endpoint_re = re.compile(r'Endpoint(?: \(\S+\))?=(\S+)')
+    ssh_re = re.compile(r'\s*SSH=(\S*):(.*)')
+    for line in lfs_env.splitlines():
+        m = endpoint_re.match(line)
+        if m:
+            if endpoint is None:
+                endpoint = m[1]
+            else:
+                break
+        m = ssh_re.match(line)
+        if m:
+            ssh_server, ssh_path = m.groups()
+            break
+    assert endpoint, f"no Endpoint= line found in git lfs env:\n{lfs_env}"
+    headers = {
+        "Content-Type": "application/vnd.git-lfs+json",
+        "Accept": "application/vnd.git-lfs+json",
+    }
+    if ssh_server:
+        ssh_command = shutil.which(os.environ.get("GIT_SSH", os.environ.get("GIT_SSH_COMMAND", "ssh")))
+        assert ssh_command, "no ssh command found"
+        with subprocess.Popen([ssh_command, ssh_server, "git-lfs-authenticate", ssh_path, "download"],
+                              stdout=subprocess.PIPE) as ssh:
+            resp = json.load(ssh.stdout)
+            assert ssh.wait() == 0, "ssh command failed"
+        endpoint = resp.get("href", endpoint)
+        for k, v in resp.get("header", {}).items():
+            headers[k.capitalize()] = v
+    url = urlparse(endpoint)
+    # this is how actions/checkout persist credentials
+    # see https://github.com/actions/checkout/blob/44c2b7a8a4ea60a981eaca3cf939b5f4305c123b/src/git-auth-helper.ts#L56-L63
+    auth = subprocess.run(["git", "config", f"http.{url.scheme}://{url.netloc}/.extraheader"], text=True,
+                          stdout=subprocess.PIPE, cwd=source_dir).stdout.strip()
+    for l in auth.splitlines():
+        k, _, v = l.partition(": ")
+        headers[k.capitalize()] = v
+    if "GITHUB_TOKEN" in os.environ:
```

Comment:
Do I read this correctly as:
0. We retrieve the lfs auth from SSH if using a SSH checkout
1. We try to retrieve the auth from `actions/checkout`
2. Then we override this by `GITHUB_TOKEN` if present

How does this deal with local clones over `http` auth? I'm a bit worried we're now getting into the game of having to support all possible ways git supports authenticating - in particular `gh` can install some auth provider for `http` links I believe.
Is there not a better way for us to always retrieve the authentication from `git`, regardless in which config option the token is stored?

Reply:
hmm that is a good point, and indeed this current system does not support repositories fetched via `http` authorized by `gh`.

The very best solution would be for `git lfs` to provide a plumbing command outputting these ephemeral URLs, so maybe we should lobby for that?

Otherwise, I'd be happy to cover at least what we semmle-code developers and CI use, without hoping to really cover all corner cases.

The standard way to retrieve authentication from git is via `git credential fill`... but that won't work in case we are using SSH or when the credentials are stored as actions does. So if after all the above we are still without an authorization header, maybe we can fall back to `git credential fill`

Reply:
> The very best solution would be for git lfs to provide a plumbing command outputting these ephemeral URLs, so maybe we should lobby for that?

Yeah that sounds like it ought to be the best. However as we also rely on user-installed `git-lfs` versions on the developer machines, the cycle of getting this in would be rather long.

Reply:
indeed :shrug: